src/dune_engine/load_rules.ml

```diff
@@ -660,6 +661,10 @@ end = struct
       Memo.return
         (Loaded.Build_under_directory_target { directory_target_ancestor })
     | Normal { rules; build_dir_only_sub_dirs; directory_targets } ->
+      let build_dir_only_sub_dirs =
+        Path.Build.Map.find build_dir_only_sub_dirs dir
```

Comment:
It probably makes sense to move this bit of logic to `Build_only_sub_dirs`. If we move a couple other helper functions there (like `singleton`) we can probably make the type abstract, which seems like a win in the long term.

Reply:
I like that idea. Did that in [cf48a7a](https://github.com/ocaml/dune/pull/7207/commits/cf48a7a982e6b427312f610b6b3598c6cbf218c9)

Reply:
Nice!